### ISSUE TITLE
Status of Saved Meta Optimizer

### ISSUE THREAD

**Issue by NoraAl** [teddykoker/learning-to-learn-jax#1]
Thanks a lot for the elegant and clean reimplementation of the paper. 

I noticed that the meta parameters saved are only from the first model in each episode. Therefore not much learning is tested.  The reasoning is as follows:
1. `meta_loss` is initialized to zero in each episode, [here](https://github.com/teddykoker/learning-to-learn-jax/blob/0a53313fc0dcaa79ca5eb3b07fe6b4d74aed4df7/train.py#L45). `meta_loss = 0`.
2. In each outer loop step, `meta_loss` accumulates the loss, [here](https://github.com/teddykoker/learning-to-learn-jax/blob/0a53313fc0dcaa79ca5eb3b07fe6b4d74aed4df7/train.py#L51). `meta_loss += loss`.
Based on that, the least and best `meta_loss` is only at the first iteration of the outer loop, or when the meta parameters are not updated. 
4. The meta parameters are saved when `meta_loss < best_loss`, [here](https://github.com/teddykoker/learning-to-learn-jax/blob/0a53313fc0dcaa79ca5eb3b07fe6b4d74aed4df7/train.py#L56). 

A quick fix could be to change the comparison from (` if meta_loss < best_loss`) to (` if loss < best_loss`).

I hope I didn't miss something. Thank you again for sharing your elegant code.

**Comment by teddykoker** [comment]
Thank you for reaching out!

I'm not sure I fully understand your reasoning, specifically:
> 2. In each outer loop step, meta_loss accumulates the loss, [here](https://github.com/teddykoker/learning-to-learn-jax/blob/0a53313fc0dcaa79ca5eb3b07fe6b4d74aed4df7/train.py#L51). `meta_loss += loss`.
Based on that, the least and best `meta_loss` is only at the first iteration of the outer loop, or when the meta parameters are not updated. 

The `meta_loss` only accumulates over the unrolls, not over the outer loop. It is set to zero at the start of each outer step, which you note in your first point. The unrolls are only performed to split the full outer step (e.g. running the optimizer for 100 epochs) into smaller parts (e.g. running the optimizer for 20 epochs, 5 times). In this case, to obtain the total loss of outer step, we simply sum the losses of each unroll.

Since `if meta_loss < best_loss` is checked after all of the unrolls are performed, it makes sure each unroll is taken into account. We want an optimizer that has the lowest total loss over the entire 100 epochs, not just a 20-epoch portion. After this check is performed, the outer loop continues to the next step where the `meta_loss` is reset to 0. This means that with each outer step, the meta optimizer can potentially achieve a lower `meta_loss`, 
and it will checkpoint the model if it does so. This can be verified by plotting `meta_loss` over time:

![loss](https://user-images.githubusercontent.com/11153048/210238390-7dc8f687-40a7-436f-ab02-dda6e661e2eb.png)

Using  `if loss < best_loss` would likely work too, but this would mean that only the last unroll would be take into account. This means that if the optimizer did very poorly for the first 80 epochs of the inner loop, but than very well for the last 20 it would still be saved. Summing the loss over all the unrolls ensures optimal performance over the entire inner loop.

Hope this all makes sense, and that I'm not missing anything either! -Teddy

**Comment by NoraAl** [comment]
Thank you for taking the time to answer. You explanation helps a lot. I may have indented the `if meta_loss < best_loss ` block. Things are clear now.

**Comment by NoraAl** [comment]
Thank you for taking the time to answer. Your explanation helps a lot. I may have indented the if `meta_loss < best_loss` block. Things are clear now.

**Comment by NoraAl** [comment]
Thank you for taking the time to answer. Your explanation helps a lot. I may have indented the `if meta_loss < best_loss` block. Things are clear now.
